goa v2 panics when using unsupported Any type

If you (improperly) use the Any type, as in the following example:
`$GOPATH/src/foo/design.go`
```package design

import (
	. "goa.design/goa/dsl"
)

var _ = Service("Something", func() {
	Method("create", func() {
		Payload(func() {
			Field(1, "name", String)
			Field(2, "bad", Any)
		})
		Result(Empty)

		GRPC(func() {
			Response(CodeOK)
		})
	})
})
```
then you get a panic without any error detail:
```$ goa gen foo/design
exit status 2
panic: cannot compute native protocol buffer type for expr.Primitive

goroutine 1 [running]:
goa.design/goa/grpc/codegen.protoBufNativeMessageTypeName(0xb301a0, 0xb232c8, 0xc0001790e8, 0x40d19f)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/protobuf.go:361 +0x2a9
goa.design/goa/grpc/codegen.protoBufMessageDef(0xc0002b6fa0, 0xc00019b2c0, 0xc0001b1746, 0x3)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/protobuf.go:239 +0x9d
goa.design/goa/grpc/codegen.protoBufMessageDef(0xc0002b6e60, 0xc00019b2c0, 0xc0001b0ad0, 0xd)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/protobuf.go:259 +0x2b1
goa.design/goa/grpc/codegen.collectMessages(0xc0002b6c80, 0xc000272ea0, 0xc0002c0270, 0x441fd1, 0xc0001ae398, 0x0)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:602 +0x7cb
goa.design/goa/grpc/codegen.ServicesData.analyze.func1(0xc0002b6c80, 0x0)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:431 +0x52
goa.design/goa/grpc/codegen.ServicesData.analyze(0xc000290330, 0xc0001957a0, 0x9)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:500 +0x2413
goa.design/goa/grpc/codegen.ServicesData.Get(0xc000290330, 0xa3ab51, 0x9, 0x40d19f)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/service_data.go:354 +0xb4
goa.design/goa/grpc/codegen.protoFile(0xc0001b819a, 0x7, 0xc0002b20c0, 0xc0001957a0, 0x1)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/proto.go:29 +0x62
goa.design/goa/grpc/codegen.ProtoFiles(0xc0001b819a, 0x7, 0xf09ba0, 0x0, 0x0, 0x0)
	/home/me/gopath/src/goa.design/goa/grpc/codegen/proto.go:23 +0xef
goa.design/goa/codegen/generator.Transport(0xc0001b819a, 0x7, 0xc0001b4e20, 0x2, 0x2, 0x3, 0x4, 0x3, 0x0, 0x0)
	/home/me/gopath/src/goa.design/goa/codegen/generator/transport.go:33 +0x5fb
goa.design/goa/codegen/generator.Generate(0x7ffcb9ae09d8, 0x1, 0xa28988, 0x3, 0x0, 0x0, 0xc00019ace0, 0x0, 0xffffffffffffffff)
	/home/me/gopath/src/goa.design/goa/codegen/generator/generate.go:62 +0x362
main.main()
	/home/me/gopath/src/foo/goa893919833/main.go:54 +0x275
```
this is hard to debug when your design is complex so perhaps do not panic but instead say what type is unsupported and the line number.

--- a/expr/grpc_endpoint.go
+++ b/expr/grpc_endpoint.go
@@ -108,6 +108,14 @@ func (e *GRPCEndpointExpr) Validate() error {
 		verr.Add(e, "Endpoint name cannot be empty")
 	}
 
+	// error if payload, result, and error type define attribute of Any type
+	// which is unsupported.
+	verr.Merge(e.hasAnyType(e.MethodExpr.Payload, "Payload"))
+	verr.Merge(e.hasAnyType(e.MethodExpr.Result, "Result"))
+	for _, er := range e.MethodExpr.Errors {
+		verr.Merge(e.hasAnyType(er.AttributeExpr, fmt.Sprintf("Error %q", er.Name)))
+	}
+
 	var hasMessage, hasMetadata bool
 	// Validate request
 	if e.Request.Type != Empty {
@@ -463,4 +471,62 @@ func getSecurityAttributes(m *MethodExpr) []string {
 		}
 	}
 	return secAttrs
+}
+
+// hasAnyType recurses through the given attribute and returns validation error
+// if any attribute is of Any type.
+func (e *GRPCEndpointExpr) hasAnyType(a *AttributeExpr, typ string, seen ...map[string]struct{}) *eval.ValidationErrors {
+	verr := new(eval.ValidationErrors)
+	if a.Type == Any {
+		verr.Add(e, "%s type is Any type which is not supported in gRPC", typ)
+	}
+	switch actual := a.Type.(type) {
+	case UserType:
+		var s map[string]struct{}
+		if len(seen) > 0 {
+			s = seen[0]
+		} else {
+			s = make(map[string]struct{})
+			seen = append(seen, s)
+		}
+		if _, ok := s[actual.ID()]; ok {
+			return verr
+		}
+		s[actual.ID()] = struct{}{}
+		verr.Merge(e.hasAnyType(actual.Attribute(), typ, seen...))
+	case *Array:
+		if IsPrimitive(actual.ElemType.Type) {
+			if actual.ElemType.Type == Any {
+				verr.Add(e, "Array element type is Any type which is not supported in gRPC")
+			}
+			return verr
+		}
+		verr.Merge(e.hasAnyType(actual.ElemType, typ, seen...))
+	case *Map:
+		if IsPrimitive(actual.KeyType.Type) {
+			if actual.KeyType.Type == Any {
+				verr.Add(e, "Map key type is Any type which is not supported in gRPC")
+			}
+		} else {
+			verr.Merge(e.hasAnyType(actual.KeyType, typ, seen...))
+		}
+		if IsPrimitive(actual.ElemType.Type) {
+			if actual.ElemType.Type == Any {
+				verr.Add(e, "Map element type is Any type which is not supported in gRPC")
+			}
+			return verr
+		}
+		verr.Merge(e.hasAnyType(actual.ElemType, typ, seen...))
+	case *Object:
+		for _, nat := range *actual {
+			if IsPrimitive(nat.Attribute.Type) {
+				if nat.Attribute.Type == Any {
+					verr.Add(e, "Attribute %q is Any type which is not supported in gRPC", nat.Name)
+				}
+				continue
+			}
+			verr.Merge(e.hasAnyType(nat.Attribute, typ, seen...))
+		}
+	}
+	return verr
 }

--- a/expr/grpc_endpoint_test.go
+++ b/expr/grpc_endpoint_test.go
@@ -1,0 +1,57 @@
+package expr_test
+
+import (
+	"testing"
+
+	"goa.design/goa/eval"
+	"goa.design/goa/expr"
+	"goa.design/goa/expr/testdata"
+)
+
+func TestGRPCEndpointValidation(t *testing.T) {
+	cases := map[string]struct {
+		DSL    func()
+		Errors []string
+	}{
+		"endpoint-with-any-type": {
+			DSL: testdata.GRPCEndpointWithAnyType,
+			Errors: []string{`service "Service" gRPC endpoint "Method": Map key type is Any type which is not supported in gRPC
+service "Service" gRPC endpoint "Method": Array element type is Any type which is not supported in gRPC
+service "Service" gRPC endpoint "Method": Attribute "invalid_primitive" is Any type which is not supported in gRPC
+service "Service" gRPC endpoint "Method": Array element type is Any type which is not supported in gRPC
+service "Service" gRPC endpoint "Method": Error "invalid_error_type" type is Any type which is not supported in gRPC
+service "Service" gRPC endpoint "Method": Map element type is Any type which is not supported in gRPC`,
+			},
+		},
+	}
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			if c.Errors == nil || len(c.Errors) == 0 {
+				expr.RunDSL(t, c.DSL)
+			} else {
+				var errors []error
+
+				err := expr.RunInvalidDSL(t, c.DSL)
+				if err != nil {
+					if merr, ok := err.(eval.MultiError); ok {
+						for _, e := range merr {
+							errors = append(errors, e.GoError)
+						}
+					} else {
+						errors = append(errors, err)
+					}
+				}
+
+				if len(c.Errors) != len(errors) {
+					t.Errorf("%s: got %d, expected the number of error values to match %d", name, len(errors), len(c.Errors))
+				} else {
+					for i, err := range errors {
+						if err.Error() != c.Errors[i] {
+							t.Errorf("%s:\ngot \t%q,\nexpected\t%q at index %d", name, err.Error(), c.Errors[i], i)
+						}
+					}
+				}
+			}
+		})
+	}
+}

--- a/expr/http_endpoint_test.go
+++ b/expr/http_endpoint_test.go
@@ -20,9 +20,9 @@ func TestHTTPRouteValidation(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {
 			if c.Error == "" {
-				expr.RunHTTPDSL(t, c.DSL)
+				expr.RunDSL(t, c.DSL)
 			} else {
-				err := expr.RunInvalidHTTPDSL(t, c.DSL)
+				err := expr.RunInvalidDSL(t, c.DSL)
 				if err.Error() != c.Error {
 					t.Errorf("got error %q, expected %q", err.Error(), c.Error)
 				}
@@ -31,7 +31,7 @@ func TestHTTPRouteValidation(t *testing.T) {
 	}
 }
 
-func TestEndpointValidation(t *testing.T) {
+func TestHTTPEndpointValidation(t *testing.T) {
 	cases := map[string]struct {
 		DSL    func()
 		Errors []string
@@ -55,11 +55,11 @@ func TestEndpointValidation(t *testing.T) {
 	for name, c := range cases {
 		t.Run(name, func(t *testing.T) {
 			if c.Errors == nil || len(c.Errors) == 0 {
-				expr.RunHTTPDSL(t, c.DSL)
+				expr.RunDSL(t, c.DSL)
 			} else {
 				var errors []error
 
-				err := expr.RunInvalidHTTPDSL(t, c.DSL)
+				err := expr.RunInvalidDSL(t, c.DSL)
 				if err != nil {
 					if merr, ok := err.(eval.MultiError); ok {
 						for _, e := range merr {
@@ -84,7 +84,7 @@ func TestEndpointValidation(t *testing.T) {
 	}
 }
 
-func TestEndpointFinalization(t *testing.T) {
+func TestHTTPEndpointFinalization(t *testing.T) {
 	cases := map[string]struct {
 		DSL          func()
 		ExpectedBody expr.DataType
@@ -100,7 +100,7 @@ func TestEndpointFinalization(t *testing.T) {
 	}
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
-			root := expr.RunHTTPDSL(t, tc.DSL)
+			root := expr.RunDSL(t, tc.DSL)
 			e := root.API.HTTP.Services[0].HTTPEndpoints[0]
 
 			if tc.ExpectedBody != nil {

--- a/expr/http_response_test.go
+++ b/expr/http_response_test.go
@@ -25,9 +25,9 @@ func TestHTTPResponseValidation(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {
 			if c.Error == "" {
-				expr.RunHTTPDSL(t, c.DSL)
+				expr.RunDSL(t, c.DSL)
 			} else {
-				err := expr.RunInvalidHTTPDSL(t, c.DSL)
+				err := expr.RunInvalidDSL(t, c.DSL)
 				if err.Error() != c.Error {
 					t.Errorf("got error %q, expected %q", err.Error(), c.Error)
 				}

--- a/expr/testdata/endpoint_dsls.go
+++ b/expr/testdata/endpoint_dsls.go
@@ -148,3 +148,26 @@ var FinalizeEndpointBodyAsPropWithExtendedTypeDSL = func() {
 		})
 	})
 }
+
+var GRPCEndpointWithAnyType = func() {
+	var Recursive = Type("Recursive", func() {
+		Field(1, "invalid_map_key", MapOf(Any, "Recursive"))
+		Field(3, "invalid_array", ArrayOf(ArrayOf(Any)))
+	})
+	var InvalidRT = ResultType("application/vnd.result", func() {
+		TypeName("RT")
+		Attributes(func() {
+			Field(1, "invalid_primitive", Any)
+			Field(2, "invalid_array", ArrayOf(Any))
+		})
+	})
+	Service("Service", func() {
+		Method("Method", func() {
+			Payload(Recursive)
+			Result(CollectionOf(InvalidRT))
+			Error("invalid_error_type", Any)
+			Error("invalid_map_type", MapOf(Int, Any))
+			GRPC(func() {})
+		})
+	})
+}

--- a/expr/testing.go
+++ b/expr/testing.go
@@ -6,9 +6,9 @@ import (
 	"goa.design/goa/eval"
 )
 
-// RunHTTPDSL returns the http DSL root resulting from running the given DSL.
+// RunDSL returns the DSL root resulting from running the given DSL.
 // Used only in tests.
-func RunHTTPDSL(t *testing.T, dsl func()) *RootExpr {
+func RunDSL(t *testing.T, dsl func()) *RootExpr {
 	setupDSLRun()
 
 	// run DSL (first pass)
@@ -25,9 +25,9 @@ func RunHTTPDSL(t *testing.T, dsl func()) *RootExpr {
 	return Root
 }
 
-// RunInvalidHTTPDSL returns the error resulting from running the given DSL.
+// RunInvalidDSL returns the error resulting from running the given DSL.
 // It is used only in tests.
-func RunInvalidHTTPDSL(t *testing.T, dsl func()) error {
+func RunInvalidDSL(t *testing.T, dsl func()) error {
 	setupDSLRun()
 
 	// run DSL (first pass)
@@ -44,25 +44,6 @@ func RunInvalidHTTPDSL(t *testing.T, dsl func()) error {
 	t.Fatal("expected a DSL evaluation error - got none")
 
 	return nil
-}
-
-// RunGRPCDSL returns the gRPC DSL root resulting from running the given DSL.
-// Used only in tests.
-func RunGRPCDSL(t *testing.T, dsl func()) *RootExpr {
-	setupDSLRun()
-
-	// run DSL (first pass)
-	if !eval.Execute(dsl, nil) {
-		t.Fatal(eval.Context.Error())
-	}
-
-	// run DSL (second pass)
-	if err := eval.RunDSL(); err != nil {
-		t.Fatal(err)
-	}
-
-	// return generated root
-	return Root
 }
 
 func setupDSLRun() {

--- a/grpc/codegen/testing.go
+++ b/grpc/codegen/testing.go
@@ -15,7 +15,7 @@ func RunGRPCDSL(t *testing.T, dsl func()) *expr.RootExpr {
 	// reset all roots and codegen data structures
 	service.Services = make(service.ServicesData)
 	GRPCServices = make(ServicesData)
-	return expr.RunGRPCDSL(t, dsl)
+	return expr.RunDSL(t, dsl)
 }
 
 func sectionCode(t *testing.T, section ...*codegen.SectionTemplate) string {

--- a/http/codegen/client_body_types_test.go
+++ b/http/codegen/client_body_types_test.go
@@ -22,7 +22,7 @@ func TestBodyTypeDecl(t *testing.T) {
 	}
 	for _, c := range cases {
 		t.Run(c.Name, func(t *testing.T) {
-			expr.RunHTTPDSL(t, c.DSL)
+			RunHTTPDSL(t, c.DSL)
 			fs := clientType(genpkg, expr.Root.API.HTTP.Services[0], make(map[string]struct{}))
 			section := fs.SectionTemplates[1]
 			code := codegen.SectionCode(t, section)

--- a/http/codegen/testing.go
+++ b/http/codegen/testing.go
@@ -13,7 +13,7 @@ func RunHTTPDSL(t *testing.T, dsl func()) *expr.RootExpr {
 	// reset all roots and codegen data structures
 	service.Services = make(service.ServicesData)
 	HTTPServices = make(ServicesData)
-	return expr.RunHTTPDSL(t, dsl)
+	return expr.RunDSL(t, dsl)
 }
 
 // makeGolden returns a file object used to write test expectations. If

